feat: implement Copy From

WenyXu opened the pull request:
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

### Parser and SqlHandler

- [x] Add `Copy From` Parser
- [x] Add `Copy From` SqlHandler

### Unit Test for `Copy From`
- [x] Import data from fs by specifying the filename
- [x]  Import data from fs by specifying the path
- [x]  Import data from fs by specifying the path with regex pattern

### Test with cloud service backends
- [ ] S3

### Support other cloud service backends and Tests
We can leave the following tasks to new contributors. It's a good starting for those who want to contribute to the project at university.

- [ ] [Google Cloud Storage](https://cloud.google.com/storage)
- [ ] [Azure Storage Blob](https://azure.microsoft.com/en-us/services/storage/blobs/)
- [ ] [Aliyun Object Storage Service](https://www.aliyun.com/product/oss) (OSS)
- [ ] [Huawei Cloud Object Storage](https://www.huaweicloud.com/intl/en-us/product/obs.html) (OBS)




### Syntax (Draft) 
```
COPY [<database>.]<table_name>
FROM { '<path>/[<filename>]' }
[ [ WITH ]
 (
   [ FORMAT =  { parquet } ]
   [ PATTERN = '<regex_pattern>' ]
 )
]
[ [ CONNECTION ]
 (
   [ ENDPOINT = '<uri>' ]
   [ ACCESS_KEY_ID = '<key_id>' ]
   [ SECRET_ACCESS_KEY = '<access_key>' ]
   [ SESSION_TOKEN = '<token>' ]
   [ REGION = '<region>' ]
   [ ENABLE_VIRTUAL_HOST_STYLE = '<boolean>']
   ..
 )
]
```

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#1038

evenyag commented:
I'd suggest putting the design drafts and the remaining task list to issue #1038 or a new tracking issue. PR isn't a suitable place to track them. PR over 1k lines is difficult to review

> ### Test with cloud service backends
> * [ ]  S3
> 
> ### Support other cloud service backends and Tests
> We can leave the following tasks to new contributors. It's a good starting for those who want to contribute to the project at university.
> 
> * [ ]  [Google Cloud Storage](https://cloud.google.com/storage)
> * [ ]  [Azure Storage Blob](https://azure.microsoft.com/en-us/services/storage/blobs/)
> * [ ]  [Aliyun Object Storage Service](https://www.aliyun.com/product/oss) (OSS)
> * [ ]  [Huawei Cloud Object Storage](https://www.huaweicloud.com/intl/en-us/product/obs.html) (OBS)
> 
> ### Syntax (Draft)
> ```
> COPY [<database>.]<table_name>
> FROM { '<path>/[<filename>]' }
> [ [ WITH ]
>  (
>    [ FORMAT =  { parquet } ]
>    [ PATTERN = '<regex_pattern>' ]
>  )
> ]
> [ [ CONNECTION ]
>  (
>    [ ENDPOINT = '<uri>' ]
>    [ ACCESS_KEY_ID = '<key_id>' ]
>    [ SECRET_ACCESS_KEY = '<access_key>' ]
>    [ SESSION_TOKEN = '<token>' ]
>    [ REGION = '<region>' ]
>    [ ENABLE_VIRTUAL_HOST_STYLE = '<boolean>']
>    ..
>  )
> ]
> ```

v0y4g3r commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,254 @@
+
+        let mut futs = Vec::new();
+
+        for column_chunk in column_chunks.iter() {
+            let columns_values = fields
+                .clone()
+                .into_iter()
+                .zip(column_chunk.iter())
+                .map(|(f, c)| (f, c.clone()))
+                .collect::<HashMap<String, VectorRef>>();
```
Simple consume the column chunks to avoid a clone
```suggestion
        for column_chunk in column_chunks.into_iter() {
            let columns_values = fields
                .iter()
                .cloned()
                .zip(column_chunk.into_iter())
                .collect();
```

v0y4g3r commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,254 @@
+            })
+            .collect::<Vec<_>>();
+
+        let mut futs = Vec::new();
+
```
```suggestion
        let mut futs = Vec::with_capacity(column_chunks.len());
```

v0y4g3r commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,254 @@
+        let column_chunks = buf
+            .iter()
+            .map(|c| {
+                // TODO: how to remove the unwrap?
```
We cannot simply unwrap here, because there are still some arrays in arrow that we don't support.

v0y4g3r commented:
I'm on leave today so I quickly went through this PR. Maybe request @evenyag @fengjiachun for a thorough review.

waynexia commented:
IMO things like parsing URI are out of the scope of this task. OpenDAL will handle it for us. E.g., when creating an S3 connection it simply requires a `&str` as the endpoint: https://docs.rs/opendal/latest/opendal/services/struct.S3.html#method.endpoint 

From my perspective, in the parser we only need to receive necessary parameters like backend (FS or S3 etc), access key or endpoint as plain strings. And maybe even don't verify its format. Let the execution layer check and report errors for us. Does this make sense to you?

evenyag reviewed (commented):
I went through the parser part quickly and left some comments about the style. I'd suggest breaking this PR into several parts to speed up review:
- implementing the parser
- implementing the SQL handler

Both the SQL syntax and the parser implementation are too complex. Could we still retrieve parameters from the `WITH` clause?
```
WITH(FORMAT='parquet', ENDPOINT_TYPE='S3', S3_ENDPOINT='')
```

Then the parser puts all the parameters in a map instead of parsing their content.

We might need to further discuss this in #1038

WenyXu commented:
> IMO things like parsing URI are out of the scope of this task. OpenDAL will handle it for us. E.g., when creating an S3 connection it simply requires a `&str` as the endpoint: https://docs.rs/opendal/latest/opendal/services/struct.S3.html#method.endpoint
> 
> From my perspective, in the parser we only need to receive necessary parameters like backend (FS or S3 etc), access key or endpoint as plain strings. And maybe even don't verify its format. Let the execution layer check and report errors for us. Does this make sense to you?


> I went through the parser part quickly and left some comments about the style. I'd suggest breaking this PR into several parts to speed up review:
> 
> * implementing the parser
> * implementing the SQL handler
> 
> Both the SQL syntax and the parser implementation are too complex. Could we still retrieve parameters from the `WITH` clause?
> 
> ```
> WITH(FORMAT='parquet', ENDPOINT_TYPE='S3', S3_ENDPOINT='')
> ```
> 
> Then the parser puts all the parameters in a map instead of parsing their content.
> 
> We might need to further discuss this in #1038

Currently, The parser will
- Distinguish protocol 
   - s3, azblob, gcs, obs (corresponding to OpenDAL's backend)
   - fs (starting with `/` or ``)
   - http/https (starting with "http")
- Extract 
   - bucket(s3) 
   - container (azblob,gcs,obs)
   - hostname (http/https)
- Split path and filename(if any)

```
s3://bucket/to/path/filename
____|______|_______|________|

```
Actually, We need the `bucket/container` parameter to build the accessor.

```rust
let mut builder = S3::default();
builder.bucket("bucket");
```

If we retrieve all parameters from the `WITH`, may be a bit verbose. it will be: 

```
COPY tab FROM /to/path/ WITH(FORMAT='parquet', ENDPOINT_TYPE='S3', S3_ENDPOINT='https:://xxx.com', BUCKET='bucket')
```

And, If we need to support importing the table from the https protocol in the future, we still need to parse the URI. (The Databend supports this feature, So I assume we also need it).

So I think we can keep retrieving the `bucket` and protocol info from the URI (e.g., 's3://bucket/to/path/' )

```
COPY tab FROM 'https://host/to/path/'  WITH(FORMAT='parquet')
COPY tab FROM 's3://bucket/to/path/'  WITH(FORMAT='parquet') CONNECTION(ENDPOINT='https:://xxx.com')
```

And I want to keep the `WITH` simple, and I introduced the `CONNECTION`,[As well as Databend did](https://databend.rs/doc/sql-commands/dml/dml-copy-into-table ). it only contains connection-related options.

I'm going to open a new issue to discuss the `Copy FROM` syntax and track the remaining tasks to support other cloud services.

WenyXu commented:
It seems like the `uris` package introduces a lot of modification of `Cargo.lock` 😢

evenyag commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,461 @@
+
+                    ObjectStore::new(accessor).finish()
+                }
+                &_ => {
```
style: You don't need to write this `&` in Rust
```suggestion
                _ => {
```


See https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html

evenyag commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,461 @@
+        })
+    }
+
+    pub async fn list(&self) -> Result<Vec<Object>> {
```
Looks like we could hold the regex in DataSource instead of Source. Then `Source` becomes much simpler as it only contains two variants.
```rust
match &self.source {
    Source::Dir => {
        let streamer = self
            .object_store
            .object("/")
            .list()
            .await
            .context(error::ListObjectsSnafu { path: &self.path })?;
        streamer
            .try_filter(|f| {
                let res = if let Some(regex) = &self.regex {
                    regex.is_match(f.name())
                } else {
                    true
                };
                future::ready(res)
            })
            .try_collect::<Vec<_>>()
            .await
            .context(error::ListObjectsSnafu { path: &self.path })
    }
    Source::Filename(filename) => {
        let obj = self.object_store.object(filename);

        Ok(vec![obj])
    }
}
```

evenyag commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,461 @@
+        pattern: Option<String>,
+        connection: HashMap<String, String>,
+    ) -> Result<DataSource> {
+        let result = SCHEMA_PATTERN.find(&uri);
```
[url](https://docs.rs/url/latest/url/) has `schema()` method to get the schema.

killme2008 commented:
@WenyXu hi, are u still working on this feature?

WenyXu commented:
> @WenyXu hi, are u still working on this feature?

Yep, I’m working on it

evenyag commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,451 @@
+use crate::error::{self, Result};
+use crate::sql::SqlHandler;
+
+lazy_static! {
+    static ref SCHEMA_PATTERN: Regex = Regex::new(r"^\w*://").unwrap();
+}
```
Unused code
```suggestion
```

evenyag commented on src/datanode/src/sql/copy_table_from.rs:
```diff
@@ -0,0 +1,451 @@
+        }
+
+        if let Some(enable_str) = connection.get(ENABLE_VIRTUAL_HOST_STYLE) {
+            let enable = enable_str.as_str().parse::<bool>();
+
+            match enable {
+                Ok(true) => {
+                    builder.enable_virtual_host_style();
+                }
+                Err(_) => {
+                    return Err(error::Error::InvalidConnection {
+                        msg: format!("failed to parse the {}", ENABLE_VIRTUAL_HOST_STYLE),
+                    })
+                }
+                _ => (),
+            }
```
nitpick(non-blocking): We could simplify the code
```suggestion
            let enable = enable_str.as_str().parse::<bool>().map_err(|e| {
                error::InvalidConnectionSnafu {
                    msg: format!("failed to parse the option {}={}, {}", ENABLE_VIRTUAL_HOST_STYLE, enable_str, e),
                }.build()
            })?;
            if enable {
                builder.enable_virtual_host_style();
            }
```

evenyag reviewed (approved):
Generally LGTM. I think we need more tests, especially the copy from S3. I think we could merge this initial implementation.

evenyag commented:
BTW, OpenDAL's root method doesn't support relative paths. We still some works to handle such case.

https://docs.rs/opendal/latest/opendal/services/struct.Fs.html#via-builder

v0y4g3r commented on src/datanode/Cargo.toml:
```diff
@@ -60,6 +61,10 @@ tokio-stream = { version = "0.1", features = ["net"] }
+regex = "1.6"
+lazy_static = "1.4"
+async-compat = "0.2"
+url = "2.3.1"
```
We still need to sort the dependency and cleanup version.

v0y4g3r reviewed (approved):
LGTM

killme2008 commented:
Great job! Thanks. ❤️
